src/riak_core_worker_pool.erl

```diff
@@ -278,3 +291,56 @@ discard_queued_work(Q, Mod) ->
             ok
     end.
 
+-spec poolboy_checkin(pid(), pid(), state()) -> {ok, state()}.
+poolboy_checkin(Pool, Worker, State) ->
```

Comment:
It is not wrong, but changing state as a side effect of a little function called makes it harder to follow the code. I kind of like to have state updates very declarative in a small scope.
In this case that would mean that you pass the pool name in as last argument, instead of the state, and return the checkout or an error. Then you do the state update in the function calling this one...

I do see that this may give some code duplication and probably it's fine this way, but it is harder to get an idea of what happens to the state in the main "loop".